.github/workflows/core.yml

```diff
@@ -33,24 +33,93 @@ permissions:
   contents: read # to fetch code (actions/checkout)
 
 jobs:
+  # ============================================
+  # Job 1: Prepare Docker image
+  # ============================================
+  prepare-python-r-env:
+    runs-on: ubuntu-24.04
+    permissions:
+      contents: read
+      packages: write
+    outputs:
+      image-name: ${{ steps.image.outputs.name }}
+      pom-hash: ${{ steps.hash.outputs.pom }}
+
+    steps:
+      - name: Checkout
+        uses: actions/checkout@v4
+
+      - name: Calculate pom.xml hash
+        id: hash
+        run: |
+          echo "pom=${{ hashFiles('**/pom.xml') }}" >> $GITHUB_OUTPUT
+
+      - name: Generate image name with hash
+        id: image
+        run: |
+          # Include both environment file AND Dockerfile in hash calculation
+          COMBINED_HASH=$(cat testing/env_python_3.9_with_R.yml .github/docker/python-r-env.Dockerfile | sha256sum | cut -d' ' -f1 | cut -c1-12)
+          IMAGE_NAME="ghcr.io/${{ github.repository_owner }}/zeppelin-test-env:py39-r-${COMBINED_HASH}"
+
+          echo "name=${IMAGE_NAME}" >> $GITHUB_OUTPUT
+
+      - name: Check if image exists
+        id: check
+        run: |
+          if docker manifest inspect ${{ steps.image.outputs.name }} >/dev/null 2>&1; then
+            echo "exists=true" >> $GITHUB_OUTPUT
+          else
+            echo "exists=false" >> $GITHUB_OUTPUT
+          fi
+
+      - name: Set up Docker Buildx
+        if: steps.check.outputs.exists != 'true'
+        uses: docker/setup-buildx-action@v3
+
+      - name: Log in to GHCR
+        if: steps.check.outputs.exists != 'true'
+        uses: docker/login-action@v3
+        with:
+          registry: ghcr.io
+          username: ${{ github.actor }}
+          password: ${{ secrets.GITHUB_TOKEN }}
+
+      - name: Build and push if needed
+        if: steps.check.outputs.exists != 'true'
+        uses: docker/build-push-action@v5
+        with:
+          context: .
+          file: .github/docker/python-r-env.Dockerfile
+          push: true
+          tags: |
+            ${{ steps.image.outputs.name }}
+            ghcr.io/${{ github.repository_owner }}/zeppelin-test-env:py39-r-latest
+          cache-from: type=gha
+          cache-to: type=gha,mode=max
```

Comment:
Is the workflow defined here the same as the one in `build-docker-image.yml`?

If so, I'd suggest converting `build-docker-image.yml` into a reusable workflow (using `on: workflow_call`).

That way, you can keep the current separation while allowing other wofkflows, such as `prepare-python-r-env`, to call it via the `uses` keyword.

For reference, here are the cases I mentioned:

- Reusable workflow being called
  - https://github.com/apache/airflow/blob/5845599cb740b1c6bce3fcb5efc9688a71f4b2a7/.github/workflows/ci-image-build.yml
- Workflow that calls another workflow
  - https://github.com/apache/airflow/blob/5845599cb740b1c6bce3fcb5efc9688a71f4b2a7/.github/workflows/ci-amd-arm.yml#L237